addons/controllers/clusterbootstrap_controller_test.go

```diff
@@ -463,6 +502,103 @@ var _ = Describe("ClusterBootstrap Reconciler", func() {
 					}, waitTimeout, pollingInterval).Should(BeTrue())
 				})
 
+				By("Test ClusterBootstrap webhook validateUpdate ", func() {
+					// fetch the latest clusterbootstrap
+					Eventually(func() bool {
+						err = k8sClient.Get(ctx, client.ObjectKeyFromObject(cluster), clusterBootstrap)
+						if err != nil {
+							return false
+						}
+						for _, ownerRef := range clusterBootstrap.OwnerReferences {
+							if ownerRef.UID == cluster.UID {
+								return true
+							}
+						}
+						return false
+					}, waitTimeout, pollingInterval).Should(BeTrue())
+
+					// CNI can't be nil
+					mutateClusterBootstrap := clusterBootstrap.DeepCopy()
+					mutateClusterBootstrap.Spec.CNI = nil
+					err := k8sClient.Update(ctx, mutateClusterBootstrap)
+					Expect(err).To(HaveOccurred())
+					Expect(strings.Contains(err.Error(), "spec.cni: Invalid value: \"null\": package can't be nil")).To(BeTrue())
+
+					// Kapp can't be nil
+					mutateClusterBootstrap = clusterBootstrap.DeepCopy()
+					mutateClusterBootstrap.Spec.Kapp = nil
+					err = k8sClient.Update(ctx, mutateClusterBootstrap)
+					Expect(err).To(HaveOccurred())
+					Expect(strings.Contains(err.Error(), "spec.kapp: Invalid value: \"null\": package can't be nil")).To(BeTrue())
+
+					// CSI can be nil
+					mutateClusterBootstrap = clusterBootstrap.DeepCopy()
+					mutateClusterBootstrap.Spec.CSI = nil
+					err = k8sClient.Update(ctx, mutateClusterBootstrap)
+					Expect(err).ToNot(HaveOccurred())
+
+					// Package CR must exist
+					mutateClusterBootstrap = clusterBootstrap.DeepCopy()
+					mutateClusterBootstrap.Spec.CNI.RefName = "antrea.tanzu.vmware.com.1.2.5--vmware.1-tkg.1"
+					err = k8sClient.Update(ctx, mutateClusterBootstrap)
+					Expect(strings.Contains(err.Error(), "\"antrea.tanzu.vmware.com.1.2.5--vmware.1-tkg.1\" not found")).To(BeTrue())
+
+					// Package Refname can't change
+					mutateClusterBootstrap = clusterBootstrap.DeepCopy()
+					mutateClusterBootstrap.Spec.CNI.RefName = "foobar1.example.com.1.17.2"
+					err = k8sClient.Update(ctx, mutateClusterBootstrap)
+					Expect(strings.Contains(err.Error(), "new package refName and old package refName should be the same")).To(BeTrue())
+
+					// Package can't be downgraded
+					mutateClusterBootstrap = clusterBootstrap.DeepCopy()
+					mutateClusterBootstrap.Spec.CNI.RefName = "antrea.tanzu.vmware.com.0.13.3--vmware.1-tkg.1"
+					err = k8sClient.Update(ctx, mutateClusterBootstrap)
+					Expect(strings.Contains(err.Error(), "package downgrade is not allowed")).To(BeTrue())
+
+					// Additional package can't be removed
+					mutateClusterBootstrap = clusterBootstrap.DeepCopy()
+					mutateClusterBootstrap.Spec.AdditionalPackages = mutateClusterBootstrap.Spec.AdditionalPackages[:len(mutateClusterBootstrap.Spec.AdditionalPackages)-1]
+					err = k8sClient.Update(ctx, mutateClusterBootstrap)
+					Expect(strings.Contains(err.Error(), "missing updated additional package")).To(BeTrue())
```

Comment:
Could we add a testcase for changing provider to secret and inline?

Reply:
Added